crates/biome_js_analyze/src/analyzers/nursery/use_as_const_assertion.rs

```diff
@@ -0,0 +1,280 @@
+use crate::JsRuleAction;
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Ast, Rule, RuleDiagnostic,
+};
+use biome_console::markup;
+use biome_diagnostics::Applicability;
+use biome_js_factory::make;
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsLiteralExpression, AnyTsName, AnyTsType, JsInitializerClause,
+    JsPropertyClassMember, JsSyntaxKind, JsVariableDeclarator, TsAsExpression,
+    TsTypeAssertionExpression,
+};
+use biome_rowan::{declare_node_union, AstNode, BatchMutationExt, TextRange, TriviaPieceKind};
+
+declare_rule! {
+    /// Enforce the use of `as const` over literal type and type annotation.
+    ///
+    /// In TypeScript, there are three common ways to specify that a value is of a specific type (like '2') and not a general type (like 'number'):
+    ///
+    /// 1. `as const`: telling TypeScript to infer the literal type automatically
+    /// 2. `as` with a literal type: explicitly telling the literal type to TypeScript
+    /// 3. angle bracket assertion: explicitly telling the literal type to TypeScript
+    /// 4. type annotation: explicitly telling the literal type to TypeScript when declare variables
+    ///
+    /// The rule suggests to use `as const` when you're using `as` with a literal type or type annotation, since `as const` is simpler and doesn't require retyping the value.
+    ///
+    /// Source: https://typescript-eslint.io/rules/prefer-as-const/
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```ts,expect_diagnostic
+    /// let bar: 2 = 2;
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// let foo = <'bar'>'bar';
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// let foo = { bar: 'baz' as 'baz' };
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```ts
+    /// let foo = 'bar';
+    /// let foo = 'bar' as const;
+    /// let foo: 'bar' = 'bar' as const;
+    /// let bar = 'bar' as string;
+    /// let foo = { bar: 'baz' };
+    /// ```
+    ///
+    pub(crate) UseAsConstAssertion {
+        version: "next",
+        name: "useAsConstAssertion",
+        recommended: true,
+    }
+}
+
+declare_node_union! {
+    pub(crate) Query = TsAsExpression | TsTypeAssertionExpression | JsVariableDeclarator | JsPropertyClassMember
```

Comment:
I think we should also handle `JsFormalParameter`.

Maybe we could match against `JsInitializerClause` instead of matching on `JsFormalParameter`, `JsVariableDeclarator`, and `JsPropertyClassMember`. Not sure about that.

Reply:
Could you give me an lint example for `JsFormalParameter`? I think that is used on function and class signature.

> Maybe we could match against JsInitializerClause instead of matching on JsFormalParameter, JsVariableDeclarator, and JsPropertyClassMember. Not sure about that.
I'll take another look at the query.

Reply:
> Could you give me an lint example for JsFormalParameter? I think that is used on function and class signature.

```js
class C {
    constrcutor(protected prop = 1 as 1) {}
}

function f(param = 1 as 1) {}
```

Although, it is not so useful.

Reply:
oh i got it, default parameter. Thanks!

Reply:
I've reworked the code and it looks like a query to `JsFormalParameter` is not necessary. We can handle the case with `TsAsExpression`. 


I tried to add `JsInitializerClause` to the Query, but ended up needing access to the node outside of it, which didn't improve the code much. So I did not change the object of the `Query`.